File: src/test/java/dev/jiaming/ai_interview/resume/ResumeStorageReconcilerTests.java
package dev.jiaming.ai_interview.resume;

import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import java.util.List;
import java.util.UUID;

import org.junit.jupiter.api.Test;

class ResumeStorageReconcilerTests {

	private final ResumePersistenceService persistenceService = mock(ResumePersistenceService.class);

	private final ResumeStorageService storageService = mock(ResumeStorageService.class);

	private final ResumeStorageReconciler reconciler = new ResumeStorageReconciler(
		persistenceService, storageService
	);

	@Test
	void retriesDeletionAndClearsKeyOnlyAfterSuccess() {
		FailedResumeStorage failed = new FailedResumeStorage(UUID.randomUUID(), "resumes/key");
		when(persistenceService.findUnappliedTerminalFailures(25)).thenReturn(List.of());
		when(persistenceService.findFailedStorageObjects(25)).thenReturn(List.of(failed));

		reconciler.deleteFailedResumeObjects();

		verify(storageService).delete(failed.storageKey());
		verify(persistenceService).clearStorageKey(failed.resumeId(), failed.storageKey());
	}

	@Test
	void failedDeletionLeavesKeyForNextRun() {
		FailedResumeStorage failed = new FailedResumeStorage(UUID.randomUUID(), "resumes/key");
		when(persistenceService.findUnappliedTerminalFailures(25)).thenReturn(List.of());
		when(persistenceService.findFailedStorageObjects(25)).thenReturn(List.of(failed));
		org.mockito.Mockito.doThrow(new IllegalStateException("offline"))
			.when(storageService).delete(failed.storageKey());

		reconciler.deleteFailedResumeObjects();

		verify(persistenceService, never()).clearStorageKey(failed.resumeId(), failed.storageKey());
	}

	@Test
	void appliesTerminalJobFailureEvenWhenWorkerCleanupWasInterrupted() {
		FailedResumeJob failedJob = new FailedResumeJob(
			UUID.randomUUID(), "RESUME_EXTRACTION_FAILED", "Malformed PDF"
		);
		when(persistenceService.findUnappliedTerminalFailures(25)).thenReturn(List.of(failedJob));
		when(persistenceService.findFailedStorageObjects(25)).thenReturn(List.of());

		reconciler.deleteFailedResumeObjects();

		verify(persistenceService).markFailed(
			failedJob.resumeId(), failedJob.errorCode(), failedJob.errorMessage()
		);
	}
}

File: src/test/kotlin/dev/jiaming/ai_interview/resume/ResumeStorageReconcilerTests.kt
package dev.jiaming.ai_interview.resume

import java.util.UUID
import org.junit.jupiter.api.Test
import org.mockito.Mockito

class ResumeStorageReconcilerTests {
	private val persistenceService = Mockito.mock(ResumePersistenceService::class.java)
	private val storageService = Mockito.mock(ResumeStorageService::class.java)
	private val reconciler = ResumeStorageReconciler(persistenceService, storageService)

	@Test
	fun retriesDeletionAndClearsKeyOnlyAfterSuccess() {
		val failed = FailedResumeStorage(UUID.randomUUID(), "resumes/key")
		Mockito.`when`(persistenceService.findUnappliedTerminalFailures(25)).thenReturn(emptyList())
		Mockito.`when`(persistenceService.findFailedStorageObjects(25)).thenReturn(listOf(failed))

		reconciler.deleteFailedResumeObjects()

		Mockito.verify(storageService).delete(failed.storageKey())
		Mockito.verify(persistenceService).clearStorageKey(failed.resumeId(), failed.storageKey())
	}

	@Test
	fun failedDeletionLeavesKeyForNextRun() {
		val failed = FailedResumeStorage(UUID.randomUUID(), "resumes/key")
		Mockito.`when`(persistenceService.findUnappliedTerminalFailures(25)).thenReturn(emptyList())
		Mockito.`when`(persistenceService.findFailedStorageObjects(25)).thenReturn(listOf(failed))
		Mockito.doThrow(IllegalStateException("offline")).`when`(storageService).delete(failed.storageKey())

		reconciler.deleteFailedResumeObjects()

		Mockito.verify(persistenceService, Mockito.never()).clearStorageKey(failed.resumeId(), failed.storageKey())
	}

	@Test
	fun appliesTerminalJobFailureEvenWhenWorkerCleanupWasInterrupted() {
		val failedJob = FailedResumeJob(UUID.randomUUID(), "RESUME_EXTRACTION_FAILED", "Malformed PDF")
		Mockito.`when`(persistenceService.findUnappliedTerminalFailures(25)).thenReturn(listOf(failedJob))
		Mockito.`when`(persistenceService.findFailedStorageObjects(25)).thenReturn(emptyList())

		reconciler.deleteFailedResumeObjects()

		Mockito.verify(persistenceService).markFailed(failedJob.resumeId(), failedJob.errorCode(), failedJob.errorMessage())
	}
}
